charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py

```diff
@@ -82,44 +115,107 @@ def __init__(self, message):
+
+        # Raise if there is no data found in the relation data bag
+        if not relation_data:
+            raise GatewayRelationDataMissingError("No data found in relation data bag.")
+
+        # Check if the relation data contains the expected attributes
+        expected_attributes = ["gateway_name", "gateway_namespace"]
+        missing_attributes = [
+            attribute for attribute in expected_attributes if attribute not in relation_data
+        ]
+        if missing_attributes:
+            raise GatewayRelationDataMissingError(f"Missing attributes: {missing_attributes}")
+
+    def get_relation_data(self) -> dict:
+        """Returns a dictionary with the Gateway information.
+
+        Raises:
+            GatewayRelationDataMissingError: if data is missing entirely or some attributes
+            GatewayRelationMissingError: if there is no related application
+        """
+        # Run pre-flight checks
+        relation = self.model.get_relation(self.relation_name)
+        self._relation_preflight_checks(relation=relation)
+
+        # Get relation data from remote app
+        relation_data = relation.data[relation.app]
 
         return {
-            "gateway_name": data["gateway_name"],
-            "gateway_namespace": data["gateway_namespace"],
+            "gateway_name": relation_data["gateway_name"],
+            "gateway_namespace": relation_data["gateway_namespace"],
         }
 
+
+class GatewayProvider(Object):
+    """Base class that represents a provider relation end.
+
+    Args:
+        provider_charm (CharmBase): the provider application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        provider_charm (CharmBase): variable for storing the provider application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, provider_charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(provider_charm, relation_name)
+        self.provider_charm = provider_charm
+        self.relation_name = relation_name
+
+    def send_gateway_relation_data(self, gateway_name: str, gateway_namespace: str) -> None:
+        """Updates the relation data bag with data from the localGateway.
+        Args:
+            gateway_name (str): the name of the Gateway the provider knows about
+            gateway_namespace(str): the namespace of the Gateway the provider knows about
+        """
+        # Update the relation data bag with localgateway information
+        relation = self.model.get_relation(self.relation_name)
+
+        # Raise if there is no related applicaton
+        if not relation:
+            raise GatewayRelationMissingError()
+
+        # Update relation data
+        relation.data[self.provider_charm.app].update(
+            {
+                "gateway_name": gateway_name,
+                "gateway_namespace": gateway_namespace,
+            }
+        )
```

Comment:
Is there a dimension missing here?  I think `relation` will be a list of data bags, one for each related application, wont it?  For example, this application is providing the gateway-info relation to two different related applications, how is that handled here?

Reply:
We see this in the old relation provider [here](https://github.com/canonical/istio-operators/blob/b2f820fff6d24d692d4135d041867e3cbff30efc/charms/istio-pilot/src/istio_gateway_info_provider.py#L15) - they loop through the relations.  

I think the same is valid on the Provider side as well, just from looking at the old Provider code [here](https://github.com/canonical/istio-operators/pull/237/files#diff-85507bb21a0ce73a75d60021f5ae8970194bd2e968a5e1fc0b954a41acae3adeL101) which had an index

Reply:
Yes, you are right. Updated.